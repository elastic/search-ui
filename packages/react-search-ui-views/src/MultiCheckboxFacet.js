import PropTypes from "prop-types";
import React from "react";

import { FacetValue } from "./types";
import { appendClassName, getFilterValueDisplay } from "./view-helpers";

function MultiCheckboxFacet({
  className,
  label,
  onMoreClick,
  onRemove,
  onSelect,
  options,
  showMore,
  showSearch,
  onSearch,
  searchPlaceholder
}) {
  return (
    <fieldset className={appendClassName("sui-facet", className)}>
      <legend className="sui-facet__title">{label}</legend>

      {showSearch && (
        <div className="sui-facet-search">
          <input
            className="sui-facet-search__text-input"
            type="search"
            placeholder={searchPlaceholder || "Search"}
            onChange={e => {
              onSearch(e.target.value);
            }}
          />
        </div>
      )}

      <div className="sui-multi-checkbox-facet">
        {options.length < 1 && <div>No matching options</div>}
        {options.map(option => {
          const checked = option.selected;
          return (
            <label
              key={`${getFilterValueDisplay(option.value)}`}
              htmlFor={`example_facet_${label}${getFilterValueDisplay(
                option.value
              )}`}
              className="sui-multi-checkbox-facet__option-label"
            >
              <div className="sui-multi-checkbox-facet__option-input-wrapper">
                <input
                  id={`example_facet_${label}${getFilterValueDisplay(
                    option.value
                  )}`}
                  type="checkbox"
                  className="sui-multi-checkbox-facet__checkbox"
                  checked={checked}
                  onChange={() =>
                    checked ? onRemove(option.value) : onSelect(option.value)
                  }
                />
                <span className="sui-multi-checkbox-facet__input-text">
                  {getFilterValueDisplay(option.value)}
                </span>
              </div>
              <span className="sui-multi-checkbox-facet__option-count">
                {option.count.toLocaleString("en")}
              </span>
            </label>
          );
        })}
      </div>

      {showMore && (
        <button
          type="button"
          className="sui-facet-view-more"
          onClick={onMoreClick}
          aria-label="Show more options"
        >
          + More
        </button>
      )}
    </fieldset>
  );
}

MultiCheckboxFacet.propTypes = {
  label: PropTypes.string.isRequired,
  onMoreClick: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired,
  onSearch: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(FacetValue).isRequired,
  showMore: PropTypes.bool.isRequired,
  className: PropTypes.string,
  showSearch: PropTypes.bool,
  searchPlaceholder: PropTypes.string
};

export default MultiCheckboxFacet;
