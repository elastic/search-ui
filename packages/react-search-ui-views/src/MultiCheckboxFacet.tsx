import React from "react";

import { appendClassName, getFilterValueDisplay } from "./view-helpers";
import { FacetViewProps } from "./types";
import type { FieldValue } from "@elastic/search-ui";

function MultiCheckboxFacet({
  className,
  label,
  onMoreClick,
  onRemove,
  onSelect,
  options,
  showMore,
  showSearch,
  onSearch,
  searchPlaceholder
}: FacetViewProps) {
  return (
    <fieldset className={appendClassName("sui-facet", className)}>
      <legend className="sui-facet__title">{label}</legend>

      {showSearch && (
        <div className="sui-facet-search">
          <input
            className="sui-facet-search__text-input"
            type="search"
            placeholder={searchPlaceholder || "Search"}
            onChange={(e) => {
              onSearch(e.target.value);
            }}
          />
        </div>
      )}

      <div className="sui-multi-checkbox-facet">
        {options.length < 1 && <div>No matching options</div>}
        {options.map((option) => {
          const checked = option.selected;
          const value = option.value as FieldValue;
          return (
            <label
              key={`${getFilterValueDisplay(option.value)}`}
              htmlFor={`example_facet_${label}${getFilterValueDisplay(
                option.value
              )}`}
              className="sui-multi-checkbox-facet__option-label"
            >
              <div className="sui-multi-checkbox-facet__option-input-wrapper">
                <input
                  data-transaction-name={`facet - ${label}`}
                  id={`example_facet_${label}${getFilterValueDisplay(
                    option.value
                  )}`}
                  type="checkbox"
                  className="sui-multi-checkbox-facet__checkbox"
                  checked={checked}
                  onChange={() => (checked ? onRemove(value) : onSelect(value))}
                />
                <span className="sui-multi-checkbox-facet__input-text">
                  {getFilterValueDisplay(option.value)}
                </span>
              </div>
              <span className="sui-multi-checkbox-facet__option-count">
                {option.count.toLocaleString("en")}
              </span>
            </label>
          );
        })}
      </div>

      {showMore && (
        <button
          type="button"
          className="sui-facet-view-more"
          onClick={onMoreClick}
          aria-label="Show more options"
        >
          + More
        </button>
      )}
    </fieldset>
  );
}

export default MultiCheckboxFacet;
