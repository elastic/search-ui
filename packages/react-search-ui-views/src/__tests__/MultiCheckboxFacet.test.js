import React from "react";
import MultiCheckboxFacet from "../MultiCheckboxFacet";
import { shallow } from "enzyme";

const params = {
  label: "A Facet",
  onMoreClick: jest.fn(),
  onRemove: jest.fn(),
  onSelect: jest.fn(),
  onSearch: jest.fn(),
  options: [
    {
      value: "fieldValue1",
      count: 10,
      selected: false
    },
    {
      value: "fieldValue2",
      count: 5,
      selected: true
    }
  ],
  showMore: true
};

const rangeOptions = [
  {
    count: 1,
    value: {
      from: 1,
      to: 10,
      name: "The first option"
    }
  },
  {
    count: 11,
    value: {
      from: 11,
      to: 20,
      name: "The second option"
    }
  }
];

it("renders", () => {
  const wrapper = shallow(<MultiCheckboxFacet {...params} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders falsey values correctly", () => {
  const wrapper = shallow(
    <MultiCheckboxFacet
      {...params}
      options={[
        {
          value: 0,
          count: 10,
          selected: false
        },
        {
          value: false,
          count: 20,
          selected: false
        },
        {
          value: "",
          count: 30,
          selected: false
        }
      ]}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it("renders range filters", () => {
  const wrapper = shallow(
    <MultiCheckboxFacet {...params} option={rangeOptions} />
  );
  expect(wrapper).toMatchSnapshot();
});

it("will render 'more' button if more param is true", () => {
  const wrapper = shallow(
    <MultiCheckboxFacet
      {...{
        ...params,
        showMore: true
      }}
    />
  );
  expect(wrapper.find(".sui-facet-view-more")).toHaveLength(1);
});

it("will render a no results message is no options are available", () => {
  const wrapper = shallow(
    <MultiCheckboxFacet
      {...{
        ...params,
        options: []
      }}
    />
  );
  expect(wrapper.find(".sui-multi-checkbox-facet").text()).toEqual(
    "No matching options"
  );
});

it("won't render 'more' button if more param is false", () => {
  const wrapper = shallow(
    <MultiCheckboxFacet
      {...{
        ...params,
        showMore: false
      }}
    />
  );
  expect(wrapper.find(".sui-multi-checkbox-facet__view-more")).toHaveLength(0);
});

describe("determining selected option", () => {
  it("will correctly determine which of the options is selected", () => {
    const wrapper = shallow(<MultiCheckboxFacet {...params} />);
    expect(
      wrapper
        .find("input")
        .at(0)
        .prop("checked")
    ).toBe(false);

    expect(
      wrapper
        .find("input")
        .at(1)
        .prop("checked")
    ).toBe(true);
  });
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const wrapper = shallow(
    <MultiCheckboxFacet className={customClassName} {...params} />
  );
  const { className } = wrapper.props();
  expect(className).toEqual("sui-facet test-class");
});

it("will render search input if `showSearch` param is true", () => {
  const wrapper = shallow(
    <MultiCheckboxFacet
      {...{
        ...params,
        showSearch: true
      }}
    />
  );

  expect(wrapper.find(".sui-facet-search")).toHaveLength(1);
});

it("won't render search input if `showSearch` param is false", () => {
  const wrapper = shallow(
    <MultiCheckboxFacet
      {...{
        ...params,
        showSearch: false
      }}
    />
  );

  expect(wrapper.find(".sui-facet-search")).toHaveLength(0);
});

it("should use the `searchPlaceholder` param as a search input placeholder", () => {
  const searchPlaceholder = "bingo";
  const wrapper = shallow(
    <MultiCheckboxFacet
      {...{
        ...params,
        showSearch: true,
        searchPlaceholder
      }}
    />
  );

  expect(
    wrapper.find(".sui-facet-search__text-input").prop("placeholder")
  ).toBe(searchPlaceholder);
});
