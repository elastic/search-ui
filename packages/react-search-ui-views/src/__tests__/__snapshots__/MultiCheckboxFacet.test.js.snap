// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<fieldset
  className="sui-facet"
>
  <legend
    className="sui-facet__title"
  >
    A Facet
  </legend>
  <div
    className="sui-multi-checkbox-facet"
  >
    <label
      className="sui-multi-checkbox-facet__option-label"
      htmlFor="example_facet_A FacetfieldValue1"
      key="fieldValue1"
    >
      <div
        className="sui-multi-checkbox-facet__option-input-wrapper"
      >
        <input
          checked={false}
          className="sui-multi-checkbox-facet__checkbox"
          id="example_facet_A FacetfieldValue1"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="sui-multi-checkbox-facet__input-text"
        >
          fieldValue1
        </span>
      </div>
      <span
        className="sui-multi-checkbox-facet__option-count"
      >
        10
      </span>
    </label>
    <label
      className="sui-multi-checkbox-facet__option-label"
      htmlFor="example_facet_A FacetfieldValue2"
      key="fieldValue2"
    >
      <div
        className="sui-multi-checkbox-facet__option-input-wrapper"
      >
        <input
          checked={true}
          className="sui-multi-checkbox-facet__checkbox"
          id="example_facet_A FacetfieldValue2"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="sui-multi-checkbox-facet__input-text"
        >
          fieldValue2
        </span>
      </div>
      <span
        className="sui-multi-checkbox-facet__option-count"
      >
        5
      </span>
    </label>
  </div>
  <button
    aria-label="Show more options"
    className="sui-facet-view-more"
    onClick={[MockFunction]}
    type="button"
  >
    + More
  </button>
</fieldset>
`;

exports[`renders falsey values correctly 1`] = `
<fieldset
  className="sui-facet"
>
  <legend
    className="sui-facet__title"
  >
    A Facet
  </legend>
  <div
    className="sui-multi-checkbox-facet"
  >
    <label
      className="sui-multi-checkbox-facet__option-label"
      htmlFor="example_facet_A Facet0"
      key="0"
    >
      <div
        className="sui-multi-checkbox-facet__option-input-wrapper"
      >
        <input
          checked={false}
          className="sui-multi-checkbox-facet__checkbox"
          id="example_facet_A Facet0"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="sui-multi-checkbox-facet__input-text"
        >
          0
        </span>
      </div>
      <span
        className="sui-multi-checkbox-facet__option-count"
      >
        10
      </span>
    </label>
    <label
      className="sui-multi-checkbox-facet__option-label"
      htmlFor="example_facet_A Facetfalse"
      key="false"
    >
      <div
        className="sui-multi-checkbox-facet__option-input-wrapper"
      >
        <input
          checked={false}
          className="sui-multi-checkbox-facet__checkbox"
          id="example_facet_A Facetfalse"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="sui-multi-checkbox-facet__input-text"
        >
          false
        </span>
      </div>
      <span
        className="sui-multi-checkbox-facet__option-count"
      >
        20
      </span>
    </label>
    <label
      className="sui-multi-checkbox-facet__option-label"
      htmlFor="example_facet_A Facet"
      key=""
    >
      <div
        className="sui-multi-checkbox-facet__option-input-wrapper"
      >
        <input
          checked={false}
          className="sui-multi-checkbox-facet__checkbox"
          id="example_facet_A Facet"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="sui-multi-checkbox-facet__input-text"
        />
      </div>
      <span
        className="sui-multi-checkbox-facet__option-count"
      >
        30
      </span>
    </label>
  </div>
  <button
    aria-label="Show more options"
    className="sui-facet-view-more"
    onClick={[MockFunction]}
    type="button"
  >
    + More
  </button>
</fieldset>
`;

exports[`renders range filters 1`] = `
<fieldset
  className="sui-facet"
>
  <legend
    className="sui-facet__title"
  >
    A Facet
  </legend>
  <div
    className="sui-multi-checkbox-facet"
  >
    <label
      className="sui-multi-checkbox-facet__option-label"
      htmlFor="example_facet_A FacetfieldValue1"
      key="fieldValue1"
    >
      <div
        className="sui-multi-checkbox-facet__option-input-wrapper"
      >
        <input
          checked={false}
          className="sui-multi-checkbox-facet__checkbox"
          id="example_facet_A FacetfieldValue1"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="sui-multi-checkbox-facet__input-text"
        >
          fieldValue1
        </span>
      </div>
      <span
        className="sui-multi-checkbox-facet__option-count"
      >
        10
      </span>
    </label>
    <label
      className="sui-multi-checkbox-facet__option-label"
      htmlFor="example_facet_A FacetfieldValue2"
      key="fieldValue2"
    >
      <div
        className="sui-multi-checkbox-facet__option-input-wrapper"
      >
        <input
          checked={true}
          className="sui-multi-checkbox-facet__checkbox"
          id="example_facet_A FacetfieldValue2"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="sui-multi-checkbox-facet__input-text"
        >
          fieldValue2
        </span>
      </div>
      <span
        className="sui-multi-checkbox-facet__option-count"
      >
        5
      </span>
    </label>
  </div>
  <button
    aria-label="Show more options"
    className="sui-facet-view-more"
    onClick={[MockFunction]}
    type="button"
  >
    + More
  </button>
</fieldset>
`;
